src/avx512bw/avx512bw-from-utf8.inl.cpp

```diff
@@ -0,0 +1,466 @@
+
+            mask_lookup = 4 * [
+                gt_0f,  # 0b0000
+                true,   # 0b0001
+                true,   # 0b0010
+                true,   # 0b0011
+                le_0f,  # 0b0100
+                false,  # 0b0101
+                false,  # 0b0110
+                false,  # 0b0111
+                false,  # 0b1000
+                false,  # 0b1001
+                false,  # 0b1010
+                false,  # 0b1011
+                false,  # 0b1100
+                false,  # 0b1101
+                false,  # 0b1110
+                false,  # 0b1111
+        ] */
+        const __m512i mask_lookup = _mm512_setr_epi64(
+            0x00000080c0c0c040,
+            0x0000000000000000,
+            0x00000080c0c0c040,
+            0x0000000000000000,
+            0x00000080c0c0c040,
+            0x0000000000000000,
+            0x00000080c0c0c040,
+            0x0000000000000000
+        );
+        const __m512i mask = _mm512_shuffle_epi8(mask_lookup, nibble0);
+
+        valid_4bytes = _mm512_mask_test_epi8_mask(_4bytes, mask, t0);
+    }
+
+    {
+        // all: marks all valid leading bytes
+        __mmask64 all = _kor_mask64(valid_2bytes, _kor_mask64(valid_3bytes, valid_4bytes));
+
+        // reset leading byte marks from input
+        valid = _kxor_mask64(valid, all);
+        valid = _kand_mask64(valid, tested_chars);
+    }
+
+    return valid == 0;
+}
+
+template <unsigned idx0, unsigned idx1, unsigned idx2, unsigned idx3>
+__m512i shuffle_epi128(__m512i v) {
+    static_assert((idx0 >= 0 && idx0 <= 3), "idx0 must be in range 0..3");
+    static_assert((idx1 >= 0 && idx1 <= 3), "idx1 must be in range 0..3");
+    static_assert((idx2 >= 0 && idx2 <= 3), "idx2 must be in range 0..3");
+    static_assert((idx3 >= 0 && idx3 <= 3), "idx3 must be in range 0..3");
+
+    constexpr unsigned shuffle = idx0 | (idx1 << 2) | (idx2 << 4) | (idx3 << 6);
+    return _mm512_shuffle_i32x4(v, v, shuffle);
+}
+
+template <unsigned idx>
+constexpr __m512i broadcast_epi128(__m512i v) {
+    return shuffle_epi128<idx, idx, idx, idx>(v);
+}
+
+__m512i rotate_by1_epi8(const __m512i input) {
+
+    // lanes order: 1, 2, 3, 0 => 0b00_11_10_01
+    const __m512i permuted = _mm512_shuffle_i32x4(input, input, 0x39);
+
+    return _mm512_alignr_epi8(permuted, input, 1);
+}
+
+template <typename OUTPUT>
+std::pair<const char*, OUTPUT*> validating_utf8_to_fixed_length(const char* str, size_t len, OUTPUT* dwords) {
+    constexpr bool UTF32 = std::is_same<OUTPUT, uint32_t>::value;
+    constexpr bool UTF16 = std::is_same<OUTPUT, char16_t>::value;
+    static_assert(UTF32 or UTF16, "output type has to be uint32_t (for UTF-32) or char16_t (for UTF-16)");
+
+    const char* ptr = str;
+    const char* end = ptr + len;
+
+    OUTPUT* output = dwords;
+
+    while (ptr + 64 < end) {
+        const __m512i utf8 = _mm512_loadu_si512((const __m512i*)ptr);
+        const __mmask64 ascii = _mm512_test_epi8_mask(utf8, v_80);
+        if (ascii == 0) {
```

Comment:
Around here, you will have 6 cycles on latency. 3 cycles for _mm512_test_epi8_mask and then 3 cycles for the ascii==0. So it is not as cheap as it looks.

Reply:
Under AVX, you'd have only 3 cycles of latency, thanks to _mm256_testz_si256